Variadic and template specializations for serialize and deserialize

Sebanisu opened the pull request:
#### Describe the issue (if no issue has been made)
Serialize and Deserialize were 2 large functions. I thought it might become hard to manage.

#### Proposed fix

This gives you a separate function for each component, for both serializing and deserializing. I grouped them by `Component` and placed them at the bottom of `SceneSerializer.cpp`. The default catch all case is deleted so if you add a new `Component` to `AllComponents` then you will get a **compile time error**. So you know what you forgot and can fix it.

https://github.com/TheCherno/Hazel/blob/73d478e0eb8fc98d21ffa2c3d802fb4edbc6a938/Hazel/src/Hazel/Scene/SceneSerializer.cpp#L234-L238

#### PR impact _(Make sure to add [closing keywords](https://help.github.com/en/articles/closing-issues-using-keywords))_
List of related issues/PRs this will solve:
I took a different approach than the previous ideas I had in https://github.com/TheCherno/Hazel/issues/496. Instead of changing `AllComponents`, I altered my function to accept more template arguments. So this way we are serializing `TagComponent`. 
https://github.com/TheCherno/Hazel/blob/73d478e0eb8fc98d21ffa2c3d802fb4edbc6a938/Hazel/src/Hazel/Scene/SceneSerializer.cpp#L515-L547

`TagComponent` is deserialized on creation of an entity. So we only need it for serialization.
https://github.com/TheCherno/Hazel/blob/73d478e0eb8fc98d21ffa2c3d802fb4edbc6a938/Hazel/src/Hazel/Scene/SceneSerializer.cpp#L208-L213



#### Additional context
Add any other context about the solution here. Did you test the solution on all (relevant) platforms?
I tested on Windows 10, with a Nvidia 980 TI.

We don't serialize with `NativeScriptComponent`. I just put a TODO there maybe we can serialize it in the future.

https://github.com/TheCherno/Hazel/blob/73d478e0eb8fc98d21ffa2c3d802fb4edbc6a938/Hazel/src/Hazel/Scene/SceneSerializer.cpp#L503-L513

CircleCollider2DComponent was missing from my original commit:
https://github.com/TheCherno/Hazel/blob/73d478e0eb8fc98d21ffa2c3d802fb4edbc6a938/Hazel/src/Hazel/Scene/SceneSerializer.cpp#L466-L501

Cherno fixed this on his stream as well. So I just had to fix the merge conflicts.
Texture wasn't being serialized https://github.com/TheCherno/Hazel/issues/561:
Added `GetPath()`
https://github.com/TheCherno/Hazel/blob/73d478e0eb8fc98d21ffa2c3d802fb4edbc6a938/Hazel/src/Hazel/Renderer/Texture.h#L17
https://github.com/TheCherno/Hazel/blob/73d478e0eb8fc98d21ffa2c3d802fb4edbc6a938/Hazel/src/Platform/OpenGL/OpenGLTexture.h#L19
Serialize TexturePath: 
https://github.com/TheCherno/Hazel/blob/73d478e0eb8fc98d21ffa2c3d802fb4edbc6a938/Hazel/src/Hazel/Scene/SceneSerializer.cpp#L340-L373

Sebanisu commented:
One other way we could of done this is the move the serialize code into member functions of each component. Then we just call that. Though This seemed more in line with how you were already doing stuff.

Drockk commented on Hazel/src/Hazel/Scene/SceneSerializer.cpp:
```diff
@@ -88,6 +88,10 @@ namespace YAML {
 
+	static void SerializeAllEntityComponents(YAML::Emitter& out, Entity entity);
+
+	static void DeseralizeAllEntryComponents(YAML::detail::iterator_value& entity, Entity& deserializedEntity);
```
Hi, you have typo in every DeseralizeAllEntryComponents it should be DeserializeAllEntryComponents

Sebanisu replied:
Good catch 👍 I will fix when I get home 🏡

Sebanisu replied:
Fixed

Sebanisu commented:
I just merged the upstream/master into my branch as it was several months old. I wanted to make sure stuff was still working. 6e309d4

I noticed `CircleCollider2DComponent` was missing from Deserialize and Serialize code. So I added that. c64ec68

`Texture` wasn't being saved.  `OpenGLTexture` does store a path. So I added a `virtual GetPath()` function. So we had something to serialize. https://github.com/TheCherno/Hazel/issues/561 c64ec68

Sebanisu commented:
I need to like cherry pick this pull request I'm really only chaning like 4 files but it's showing all these other files. It's all because I merged in upstream changes. I'm not sure what the best way to do it is....

Sebanisu commented:
This pull request was originally on dev and now on it's master. Some of the changes in my pull request are from dev.
`entt.hpp` `Entity.h` are included but I haven't touched those files. TheCherno made those commits.

There was one commit I accidentally reverted. So I had a revert revert. [7b19ce5](https://github.com/TheCherno/Hazel/pull/498/commits/7b19ce58f8394e85b95fa81e03002636fb373dc4)

Sebanisu commented:
I cherry picked my commits and reset this branch so I no longer had anything from the dev branch in my pull request.

Sebanisu commented:
Cherno added TexturePath and Tiling Factor.
Main thing I did differently is I checked to see if the path exists on saving.
I resolved the merge conflicts from the latest stream.
TODO I need to pull the code down and make sure it still builds. I thought I was done and deleted my files heh.
